source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc

```diff
@@ -0,0 +1,322 @@
+  return handleString(static_cast<std::string>(value));
+}
+
+template <typename NumberType> FilterStatus TrieMatchHandler::numberValue(NumberType value) {
+  ASSERT(last_field_id_.has_value());
+  ENVOY_LOG(trace, "TrieMatchHandler numberValue id:{} value:{}", last_field_id_.value(), value);
+  return handleString(std::to_string(value));
+}
+
+FilterStatus TrieMatchHandler::handleString(std::string value) {
+  ASSERT(node_);
+  ASSERT(last_field_id_.has_value());
+  if (steps_ == 0 && node_->children_.find(last_field_id_.value()) != node_->children_.end() &&
+      !node_->children_[last_field_id_.value()]->rule_ids_.empty()) {
+    auto on_present_node = node_->children_[last_field_id_.value()];
+    ENVOY_LOG(trace, "name: {}", on_present_node->name_);
+    parent_.handleOnPresent(std::move(value), on_present_node->rule_ids_);
+  }
+  return FilterStatus::Continue;
+}
+
+PayloadToMetadataFilter::PayloadToMetadataFilter(const ConfigSharedPtr config) : config_(config) {}
+
+void PayloadToMetadataFilter::handleOnPresent(std::string&& value,
+                                              const std::vector<uint16_t>& rule_ids) {
+  for (uint16_t rule_id : rule_ids) {
+    if (matched_rule_ids_.find(rule_id) == matched_rule_ids_.end()) {
+      return;
+    }
+    ENVOY_LOG(trace, "handleOnPresent rule_id {}", rule_id);
+
+    matched_rule_ids_.erase(rule_id);
+    ASSERT(rule_id < config_->requestRules().size());
+    const Rule& rule = config_->requestRules()[rule_id];
+    if (!value.empty() && rule.rule().has_on_present()) {
+      // We can *not* always std::move(value) here since we need `value` if multiple rules are
+      // matched. Optimize the most common usage, which is one rule per payload field.
+      if (rule_ids.size() == 1) {
+        applyKeyValue(std::move(value), rule, rule.rule().on_present());
+        break;
+      } else {
+        applyKeyValue(value, rule, rule.rule().on_present());
+      }
+    }
+  }
+}
+
+void PayloadToMetadataFilter::handleOnMissing() {
+  ENVOY_LOG(trace, "{} rules missing", matched_rule_ids_.size());
+
+  for (uint16_t rule_id : matched_rule_ids_) {
+    ENVOY_LOG(trace, "handling on_missing rule_id {}", rule_id);
+
+    ASSERT(rule_id < config_->requestRules().size());
+    const Rule& rule = config_->requestRules()[rule_id];
+    if (!rule.rule().has_on_missing()) {
+      continue;
+    }
+    applyKeyValue("", rule, rule.rule().on_missing());
+  }
+}
+
+const std::string& PayloadToMetadataFilter::decideNamespace(const std::string& nspace) const {
+  static const std::string& payloadToMetadata = "envoy.filters.thrift.payload_to_metadata";
+  return nspace.empty() ? payloadToMetadata : nspace;
+}
+
+bool PayloadToMetadataFilter::addMetadata(const std::string& meta_namespace, const std::string& key,
+                                          std::string value, ValueType type) {
+  ProtobufWkt::Value val;
+  ASSERT(!value.empty());
+
+  if (value.size() >= MAX_PAYLOAD_VALUE_LEN) {
```

Comment:
I couldn't find this documented anywhere, and I think it should be. Also consider making the the log message error-level instead of debug.

Reply:
thanks! i doc it and make the log error